### PR TITLE
Add thread termination for real-time and note generation cancellation

### PR DESCRIPTION
**Pull request by ItsSimko**
## Summary by Sourcery

Add functionality to terminate threads for real-time audio processing and note generation, enabling cancellation of these operations. Enhance cancellation handling by introducing local cancel flags to manage cancellation requests more effectively.

New Features:
- Introduce thread termination functionality for real-time audio processing and note generation, allowing users to cancel these operations.

Enhancements:
- Implement local cancel flags to improve the handling of cancellation requests in real-time audio processing. Prevents the reset if user starts recording to quickly after again.

## Summary by Sourcery

Add thread termination functionality for real-time audio processing and note generation, enabling users to cancel these operations. Enhance cancellation handling by introducing local cancel flags to manage cancellation requests more effectively.

New Features:
- Introduce thread termination functionality for real-time audio processing and note generation, allowing users to cancel these operations.

Enhancements:
- Implement local cancel flags to improve the handling of cancellation requests in real-time audio processing.

### CODE REVIEW

File: src/FreeScribe.client/client.py [hunk 1]
```diff
@@ -560,9 +571,35 @@ def send_audio_to_server():
                     os.remove(file_to_send)
                 loading_window.destroy()
 
-    if is_audio_processing_whole_canceled:
-        # reset the state of the flag
-        is_audio_processing_whole_canceled = False
+def kill_thread(thread_id):
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py:574]: This is the only way to kill the transcribing process from what I was reading as it is a blocking call so it won't recheck the flag until its done transcribing. But we want the transcribe to end and release system resources on cancel press and not finish.

[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py:574]: From testing multiprocessing causes hallucenations and increases transcribe time signifcantly

[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py:574]: There is one other solution https://github.com/openai/whisper/discussions/1841
Although this one would require us to manually modify the whisper code and then bundle it ourselves for the installer.

File: src/FreeScribe.client/client.py [hunk 2]
```diff
@@ -198,6 +199,7 @@ def is_silent(data, threshold=0.01):
 
 def realtime_text():
     global frames, is_realtimeactive, audio_queue, is_audio_processing_realtime_canceled
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py:201]: Need to look into it further. Will do.

[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py:201]: To be fair the race conditions shouldn't happen as the thread gets killed instantly on cancel. Then gets reset later on to start recording. The others are just kind of residual code I can most likely remove just needs further testing.

[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py]: Changed over in this commit: https://github.com/ClinicianFOCUS/FreeScribe/pull/142/commits/21b68d6adc71bf47a98d97ea36888a6408524b11

**Comment by ItsSimko** [comment]
@sourcery-ai review

File: src/FreeScribe.client/client.py [hunk 3]
```diff
@@ -79,12 +80,16 @@
 RATE = 16000
 
 # Application flags
-is_audio_processing_realtime_canceled = False
-is_audio_processing_whole_canceled = False
+is_audio_processing_realtime_canceled = threading.Event()
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py:83]: This is only useful if we have threads that need to exit gracefully. But for our whisper and LLM threads we have to kill the using ctypes to immediately stop the block and release resources. So the kill_thread function is easier to use.

File: src/FreeScribe.client/client.py [hunk 4]
```diff
@@ -560,9 +614,35 @@ def send_audio_to_server():
                     os.remove(file_to_send)
                 loading_window.destroy()
 
-    if is_audio_processing_whole_canceled:
-        # reset the state of the flag
-        is_audio_processing_whole_canceled = False
+def kill_thread(thread_id):
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/FreeScribe.client/client.py:617]: Yes, given what is being done I can deal with this.

**Review by yingbull — APPROVED** [review]

**Pull request merged**